Adds AWS Budgets support to the configuration file

danielslbsb opened the pull request:
### Description of changes
* This adds the necessary budget fields schema and configuration classes, and creates a budget builder file, for the Cfn template.
* Added basic field validation to the schema and unit tests to ensure they are working correctly.

### Tests
* Manual testing (creating clusters and updating configuration after creation)
  * Creating cluster without budget and adding it after creation
  * Creating cluster with a budget, then changing budget category / properties
  * Creating a cluster with a budget, then deleting the budget
* Unit test to check basic validation of the budget schema

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Drafting schema for the budget configuration

This adds the necessary budget fields, except for the CostFilters, since it is not yet clear what the structure for those will be.

Commit:
Adds budget_builder.py

Creates the Cfn template for the list of budgets in the configuration. Still need to add some parameters and match the schema/config to the class.

Commit:
Bug fixes and simplifying budget configuration

Commit:
Add unit test for the budget schema validators

Wrote unit tests for the basic validation of the budget schema, such as required fields and fields with limited options. Still need to add and test more complex validation to the cluster_config file.

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
+class BudgetSchema(BaseSchema):
+    """Represent the schema of an aws budget."""
+
+    budget_category = fields.Str(
```
We can add a validation in the schema level to validate budget_category = cluster and queue_name fields are not coexist, something like:
```
    @validates_schema
    def no_coexist_....
```
An example from validate the scheduler fields:
```
    @validates_schema
    def no_coexist_schedulers(self, data, **kwargs):
```

danielslbsb replied:
Good idea. Done!

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
+    )
+    queue_name = fields.Str(metadata={"update_policy": UpdatePolicy.SUPPORTED})
+
+    cost_filters = fields.Dict(metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
cost filter fields can only be coexist with BudgetCategory: custom, let's add a schema validation

danielslbsb replied:
Good call. Added!

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
+
+    notification = fields.Nested(
+        BudgetNotificationSchema,
+        equired=True,
```
Typo `equired` here?

danielslbsb replied:
Fixed!

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
+    )
+    queue_name = fields.Str(metadata={"update_policy": UpdatePolicy.SUPPORTED})
+
+    cost_filters = fields.Dict(metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Let's add a schema validation for the supported keys in cost_filters":

Here's an example of validation metadata dict
```
    @validates("metadata")
    def validate_metadata(self, value):
        """Validate metadata contains fieds 'name' and 'version'."""
        for key in ["Name", "Version"]:
            if key not in value.keys():
                raise ValidationError(f"{key} is required for scheduler plugin Metadata.")
```

danielslbsb replied:
Done!

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
+class BudgetLimitSchema(BaseSchema):
+    """Represent the schema of the BudgetLimit field of a budget."""
+
+    amount = fields.Int(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Is this amount require to be integer? What about float?

danielslbsb replied:
Good call, it's actually a float. Updated!

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
+    """Represent the schema of the BudgetLimit field of a budget."""
+
+    amount = fields.Int(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
+    unit = fields.Str(metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Do we know how many Unit and what Unit are supported here?

danielslbsb replied:
Not really, here's the documentation on that: https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_budgets_Spend.html

danielslbsb replied:
From testing it, it seems that the only currently accepted one is USD, so I added a check for that.

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
+        metadata={"update_policy": UpdatePolicy.SUPPORTED},
+        validate=validate.OneOf(["GREATER_THAN", "EQUAL_TO", "LESS_THAN"]),
+    )
+    threshold = fields.Int(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Does `threshold` require to be an interger? Can it be float?

danielslbsb replied:
It should be a float. Updated!

chenwany commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -856,6 +856,71 @@ def __init__(self, head_node_bootstrap_timeout: int = None, compute_node_bootstr
+        self,
+        notification_type: str = None,
+        comparison_operator: str = None,
+        threshold: int = None,
```
Please double check the threshold type here

danielslbsb replied:
Done!

chenwany commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -856,6 +856,71 @@ def __init__(self, head_node_bootstrap_timeout: int = None, compute_node_bootstr
+class BudgetLimit(Resource):
+    """Represent the configuration of a budget limit."""
+
+    def __init__(self, amount: int = None, unit: str = None):
```
Please double check the amount type here

danielslbsb replied:
Done!

chenwany commented on cli/src/pcluster/templates/cluster_stack.py:
```diff
@@ -289,6 +290,11 @@ def _add_resources(self):
+        # Budgets
+        if self.config.dev_settings and self.config.dev_settings.budgets:
+            cost_budgets = CostBudgets(self, self.config)
+            self.budget_list = cost_budgets.create_budgets()
```
Shall we move the create_budgets function into budget_builder.py? 
Add the end of the _init_ function, we can add the resources, as what we do for other builder stack

danielslbsb replied:
Good call. Done!

chenwany commented on cli/src/pcluster/templates/budget_builder.py:
```diff
@@ -0,0 +1,71 @@
+        for index, budget in enumerate(budget_list):
+
+            budget_data = budgets.CfnBudget.BudgetDataProperty(
+                budget_name=f"CfnBudget{self.cluster_config.cluster_name}" + str(index),
```
If the budget name only contains cluster_name and index, we can not recognize whether the budget is queue budget, custom budget or cluster budget from its name. Let's make the budget_name more clear

danielslbsb replied:
Good idea. I made it more descriptive

chenwany commented on cli/src/pcluster/templates/budget_builder.py:
```diff
@@ -0,0 +1,71 @@
+                        )
+                    )
+
+            budget_id = f"CfnBudget{self.cluster_config.cluster_name}" + str(index)
```
same comment for budget_id

danielslbsb replied:
Updated!

chenwany commented on cli/src/pcluster/templates/budget_builder.py:
```diff
@@ -0,0 +1,71 @@
+        self.cluster_config = cluster_config
+        self.scope = scope
+
+    def create_budgets(self):
```
The `create_budgets` function is long and contains many things, let's break it into several small functions.
First, let's rename it to _add_resources(), as we do for other builder stack files.
Then you can think about breaking the function into:
Add budget_data, add notifications_with_subscribers...
Or think about breaking it into: add_cluster_budget, add_queue_budget, add_custom_budget...

danielslbsb replied:
Good call! Restructured it.

chenwany commented:
Good job! Let's also add these fields into
example_configs awsbatch.full.yaml, slurm.full.yaml and scheduler_plugin.full.yaml

chenwany commented:
In the PR description, could you list the manually test cases you perform?

chenwany commented:
Action item: add validators for the fields and integration test in another PR

Commit:
Code review updates

Addressed comments from the code review, added unit tests to cover new validation, and fixed a small bug when updating the cluster.

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,157 @@ def make_resource(self, data, **kwargs):
+        metadata={"update_policy": UpdatePolicy.SUPPORTED},
+        validate=validate.OneOf(["GREATER_THAN", "EQUAL_TO", "LESS_THAN"]),
+    )
+    threshold = fields.Float(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
nit: we can add `validate=validate.Range(min=0)`

danielslbsb replied:
Added!

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,157 @@ def make_resource(self, data, **kwargs):
+    """Represent the schema of the BudgetLimit field of a budget."""
+
+    amount = fields.Float(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
+    unit = fields.Str(metadata={"update_policy": UpdatePolicy.SUPPORTED}, validate=validate.OneOf(["USD"]))
```
Do we want to restrict unit to only be USD?

danielslbsb replied:
From testing it, it seems that it is the only currently accepted value.

chenwany commented on cli/src/pcluster/models/cluster.py:
```diff
@@ -435,7 +435,7 @@ def _validate_and_parse_config(self, validator_suppressors, validation_failure_l
 
     @staticmethod
     def _load_additional_instance_type_data(cluster_config_dict):
-        if "DevSettings" in cluster_config_dict:
```
Let change it to something like:
```
        if "DevSettings" in cluster_config_dict and cluster_config_dict.get("DevSettings"):
            instance_types_data = cluster_config_dict.get("DevSettings").get("InstanceTypesData")
```

danielslbsb replied:
Updated! Simplified it a bit.

chenwany commented on cli/src/pcluster/templates/cluster_stack.py:
```diff
@@ -289,6 +290,10 @@ def _add_resources(self):
 
+        # Budgets
+        if self.config.dev_settings and self.config.dev_settings.budgets:
+            CostBudgets(self, self.config)
```
self.budgets=CostBudgets(self, self.config)

danielslbsb replied:
Fixed!

chenwany commented on cli/src/pcluster/templates/budget_builder.py:
```diff
@@ -0,0 +1,79 @@
+    def _add_parameters(self, budget, index):
+        """Add each budget's parameters."""
+        budget_data = budgets.CfnBudget.BudgetDataProperty(
+            budget_name=f"CfnBudget{str(index)}{self.cluster_config.cluster_name}{budget.budget_category}",
```
Let's remove CfnBudget from the name , the name can be {cluster-name}-{budget-category}-{region}

danielslbsb replied:
Improved naming!

chenwany commented on cli/src/pcluster/templates/budget_builder.py:
```diff
@@ -0,0 +1,79 @@
+        return cfn_budget
+
+
+def add_budget_notifications(budget):
```
This function is returning either None or a list. It is in general better to have the function return single type of data.
Instead you can move the if condition out of the function:
```
notifications_with_subscribers = add_budget_notifications(budget) if budget.notifications_with_subscribers is not None else None
```
Then in the function you ca remove if

danielslbsb replied:
Updated!

chenwany commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -925,13 +930,161 @@ def make_resource(self, data, **kwargs):
+        budget_category = data.get("budget_category")
+        cost_filters = data.get("cost_filters")
+        queue_name = data.get("queue_name")
+        if budget_category != "custom" and cost_filters is not None:
```
can we remove is not None here?
Can we use if else here?
For example:
```
if budget_category != "custom" and cost_filters:
    raise ValidationError("CostFilters can not be configured when BudgetCategory is set to Custom.")
elif budget_category == "custom" and not cost_filters:
    raise ValidationError("CostFilters is required when BudgetCategory is set to Custom.")
if budget_category != "queue" and not queue_name:
    raise ValidationError("QueueName can only be specified when BudgetCategory is set to Queue.")
elif budget_category == "queue" and queue_name is None:
    raise ValidationError("Queue budgets require QueueName is required when BudgetCategory is set to Queue.")
```

danielslbsb replied:
Good suggestion. Updated!

chenwany commented on cli/src/pcluster/templates/budget_builder.py:
```diff
@@ -0,0 +1,84 @@
+        if budget.budget_category == "queue":
+            name = f"{self.cluster_config.cluster_name}-{index}-{budget.queue_name} {self.cluster_config.region}"
+        elif budget.budget_category == "cluster":
+            name = f"{self.cluster_config.cluster_name}-{index} {self.cluster_config.region}"
```
Missing `-` between {index} and region?

danielslbsb replied:
Updated!

chenwany commented on cli/src/pcluster/templates/budget_builder.py:
```diff
@@ -0,0 +1,84 @@
+        elif budget.budget_category == "cluster":
+            name = f"{self.cluster_config.cluster_name}-{index} {self.cluster_config.region}"
+        else:
+            name = f"{self.cluster_config.cluster_name}-{index}-custom {self.cluster_config.region}"
```
Add `-` between custom and region?

danielslbsb replied:
Updated!

chenwany reviewed (approved):
Well done!
